### ISSUE TITLE
Asynchronous image loading via the "imageLoadFunction" option is broken for image layers

### ISSUE THREAD

**Issue by igorbremec** [openlayers/openlayers#15307]
The ImageWMS, ImageArcGISRest, ImageMapGuide and ImageStatic source classes have the `imageLoadFunction` constructor option which is supposed to facilitate alternative ways of loading image data. The documenation for the LoadFunction API doc states:

`Providing a custom imageLoadFunction can be useful to load images with post requests or - in general - through XHR requests, where the src of the image element would be set to a data URI when the content is loaded.`

However, the current implementation (v8.1.0) of the image sources makes it impossible to asynchronously load the image data. Please see the code below:

https://github.com/openlayers/openlayers/blob/f68dbccc35151351070771beca36570fcb193965/src/ol/source/ImageWMS.js#L215-L219

The `imageLoadFunction_` is expected to be a synchronous function and the `decode` function is called right after it, which will of course fail to decode image data that is not yet loaded.

I have a use case where I need to load image data asynchronously and my only option at this moment is to override the `ImageWMS` class and fix the `getImageInternal` method implementation.

My suggestion to fix this is the following:

1) Change the signature of the `LoadFunction` to: `function(image: Image, src: string): Promise<void> | void`. This doesn't break compatibility with any current implementations.
2) Change the `load` function passed to `createLoader` to something similar to the following:
```
load: (image, src) => {
  this.image.setImage(image);

  const retVal = this.imageLoadFunction_(this.image, src);
  if (retVal instanceof Promise) {
    return retVal.then(() => decode(image));
  }

  return decode(image);
}
```

**Comment by ahocevar** [comment]
This issue is a duplicate of #15097 and has been fixed already with #15103.

**Comment by igorbremec** [comment]
@ahocevar Would you please care to explain how exactly #15103 solves my problem? The fix in that PR will not work for my use case since `image.complete` will be false and `image.width` will be 0. I'm loading the image data as a blob via an AJAX request and and a URL to the blob object will be assigned to `image.src` only after the load is completed. For this use case the code will still throw an error.

**Comment by ahocevar** [comment]
@igorbremec Here is an example to test: https://codesandbox.io/s/wms-image-forked-gqtzz3?file=/main.js. If you set the version of `ol` to `8.1.0`, it won't work, if you set it back to the latest dev version, it will work. The load function I added there just waits for 3 seconds before setting the src, to simulate your async use case:
```js
imageLoadFunction(image, src) {
        setTimeout(() => {
          image.getImage().src = src;
        }, 3000);
      },
```

**Comment by igorbremec** [comment]
> @igorbremec Here is an example to test: https://codesandbox.io/s/wms-image-forked-gqtzz3?file=/main.js. If you set the version of `ol` to `8.1.0`, it won't work, if you set it back to the latest dev version, it will work. The load function I added there just waits for 3 seconds before setting the src, to simulate your async use case:
> 
> ```js
>       imageLoadFunction(image, src) {
>         setTimeout(() => {
>           image.getImage().src = src;
>         }, 3000);
>       },
> ```

Thank you. I took a deeper look at the code and realized that the load() function in ol/Image.js ends up being called. However, it still didn't work when i updateded just the Image.js so I must be missing some other changes. I'll give a try again when the next OL version is released.

**Comment by igorbremec** [comment]
> @igorbremec If you're using npm, you can `npm install ol@dev` to get the latest.

I can confirm it works on the latest dev version. Thanks again 👍
